src/rust/engine/process_execution/docker/src/docker.rs

```diff
@@ -396,146 +397,152 @@ impl process_execution::CommandRunner for CommandRunner<'_> {
         _workunit: &mut RunningWorkunit,
         req: Process,
     ) -> Result<FallibleProcessResultWithPlatform, ProcessError> {
-        let req_debug_repr = format!("{req:#?}");
-        in_workunit!(
-            "run_local_process_via_docker",
-            req.level,
-            // NB: See engine::nodes::NodeKey::workunit_level for more information on why this workunit
-            // renders at the Process's level.
-            desc = Some(req.description.clone()),
-            |workunit| async move {
-                let mut workdir = create_sandbox(
-                    self.executor.clone(),
-                    &self.work_dir_base,
-                    &req.description,
-                    self.keep_sandboxes,
-                )?;
-
-                // Obtain ID of the base container in which to run the execution for this process.
-                let (container_id, named_caches) = {
-                    let ProcessExecutionStrategy::Docker(image) =
-                        &req.execution_environment.strategy
-                    else {
-                        return Err(ProcessError::Unclassified(
-                            "The Docker execution strategy was not set on the Process, but \
-                 the Docker CommandRunner was used."
+        let mut original_error: Option<String> = None;
+        loop {
+            // Make a mutable copy of the process in the closure
+            let mut mreq = req.clone();
+            let context = context.clone();
+            // TODO: do we run loop within closure or outside of it?
```

Comment:
Pulling the question from https://github.com/pantsbuild/pants/issues/18209#issuecomment-2761458772

> What is a workunit? Should the retry attempts be within the same workunit or separate workunits? If separate, should I just clone the context or would there be any updates to make?

I believe a workunit is a piece of "tracing"-style metadata tracking, i.e. a span of time with attached metadata (like parent/children spans, and other things). This metadata is exposed to users via the `pants.backend.experimental.tools.workunit_logger` backend.

So, I think this question is "is it worth exposing separate spans for each attempt at a docker invocation". My inclination is that what you have here is fine/good, i.e. loop outside and separate workunits for each invocation.